### PR TITLE
Check dir_mode recursively in file.directory

### LINKED ISSUE #52173: file.directory fails to check directory permissions correctly
### Description of Issue/Question
Running the `file.directory` state with `test=true` always reports changes to be made, even when the directories are in the desired state.

### Setup
```
salt-log-permissions:
  file.directory:
    - name: /var/log/salt
    - recurse:
      - mode
    - dir_mode: '2770'
    - file_mode: '0660'
```

```
# salt-call state.apply salt.logs test=false
local:
----------
          ID: salt-log-permissions
    Function: file.directory
        Name: /var/log/salt
      Result: True
     Comment: Directory /var/log/salt is in the correct state
     Started: 01:49:00.152027
    Duration: 59.972 ms
     Changes:   

Summary for local
------------
Succeeded: 1
Failed:    0
------------
Total states run:     1
Total run time:  59.972 ms

# salt-call state.apply salt.logs test=true
local:
----------
          ID: salt-log-permissions
    Function: file.directory
        Name: /var/log/salt
      Result: None
     Comment: The following files will be changed:
              /var/log/salt/api: mode - 0660
              /var/log/salt/master: mode - 0660
              /var/log/salt/key: mode - 0660
              /var/log/salt/minion: mode - 0660
              /var/log/salt: mode - 0660
     Started: 01:49:13.021802
    Duration: 33.664 ms
     Changes:   

Summary for local
------------
Succeeded: 1 (unchanged=1)
Failed:    0
------------
Total states run:     1
Total run time:  33.664 ms
```

### Versions Report
```
Salt Version:
           Salt: 2019.2.0
 
Dependency Versions:
           cffi: 1.11.5
       cherrypy: unknown
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.10
        libgit2: 0.27.8
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.6.1
   mysql-python: Not Installed
      pycparser: 2.18
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: 0.27.4
         Python: 3.6.8 (default, Feb  8 2019, 21:14:59)
   python-gnupg: Not Installed
         PyYAML: 3.13
          PyZMQ: 18.0.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.3.1
 
System Versions:
           dist:   
         locale: UTF-8
        machine: amd64
        release: 11.2-STABLE
         system: FreeBSD
        version: Not Installed
```

### PR DESCRIPTION
**Pull request by amendlik**
### What does this PR do?
Fixes a problem in `states.file` where directory permissions are not checked properly with `test=true`

### What issues does this PR fix or reference?
Partially fixes #52173

### Tests written?
No

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by amendlik** [comment]
Please tell me me what needs to be done to get this change merged.

**Comment by twangboy** [comment]
@amendlik We need to get some tests written for this

File: salt/states/file.py [hunk 1]
```diff
@@ -747,18 +748,18 @@ def _check_directory(name,
                         fchange['user'] = user
                     if group is not None and group != stats.get('group'):
                         fchange['group'] = group
-                    if mode is not None and mode != stats.get('mode'):
-                        fchange['mode'] = mode
+                    if file_mode is not None and file_mode != stats.get('mode'):
```
Comment (Akm0d) [inline salt/states/file.py]: the "mode" kwarg got changed to "dir_mode", so shouldn't this be "dir_mode" instead of "file_mode"?

**Comment by amendlik** [comment]
The kwargs were confusing before this change. There was a 'mode' kwarg,
which specified the directory mode, and a 'file_mode' kwarg which specified
the file mode. To make matters worse, the section that set the file mode
shadowed the 'mode' variable, overwriting it's value and causing the next
section, which sets the directory mode, to misbehave. The section of the
file you have referenced is where the file mode is set. it relies on the
'file_mode' kwarg, which did not change.

**Review by Akm0d — CHANGES_REQUESTED** [review]
This looks good and works per my own manual test, before merging we really do need to add a test case.

**Comment by amendlik** [comment]
@Akm0d, are you going to create the test case, or are you waiting for me to do that?

**Comment by amendlik** [comment]
@Akm0d, are you going to create the test case, or are you waiting for me to do that?

**Comment by amendlik** [comment]
Third try for a response: @Akm0d are you going to create the test case, or are you waiting for me to do that?

**Comment by Akm0d** [comment]
I was waiting for you to create the test case.  Sorry it took me that long to get back to you... I didn't realize I had implied that I was going to write a test, as a reviewer I was re-affirming what @twangboy had said needed to be done for this to be merged.

**Comment by KChandrashekhar** [comment]
@amendlik Thanks for this PR. Also, to ensure this gets into 2019.2.1 release, can you reopen this in 2019.2.1 branch ?

**Comment by bryceml** [comment]
re-run docs

**Comment by Akm0d** [comment]
We will write the test :+1:

**Comment by Akm0d** [comment]
@amendlik, @twangboy and I wrote a test case for this, how do you think it looks?  It is currently failing because I think our options for  filestate._check_directory()" are incorrect.  I was hoping we could get some input/advice from your end so we can have this PR ready for merge.

File: tests/unit/states/test_file.py [hunk 2]
```diff
@@ -2028,3 +2033,67 @@ def test__bad_input(self):
...
+                os.path.join(root_tmp_dir, 'dir002', 'file_000.txt'),
+                os.path.join(root_tmp_dir, 'dir000', 'dir001', 'file_002.txt'),
+                os.path.join(root_tmp_dir, 'dir001', 'dir002'),
+                os.path.join(root_tmp_dir, 'dir002', 'dir000'),
+                os.path.join(root_tmp_dir, 'dir001'),
+            }
+            for c in changed_files:
+                os.chmod(c, 0o770)
+
+            ret = filestate._check_directory(root_tmp_dir, dir_mode=0o777, file_mode=0o777, recurse=['mode'])
```
Comment (Akm0d) [inline tests/unit/states/test_file.py]: I'm sure the test is failing because the options here are incorrect, what should we be putting in for options?

**Comment by Akm0d** [comment]
I also accidentally did an auto-format of test_file.py from pycharm ... a lot of the indentation was really bad but idk if it's kosher to mix that into another PR or if cosmetic changes are allowed, should I go back and fix that @twangboy?

**Comment by Akm0d** [comment]
Cherry-picked these changes to the 2019.2.1 branch and created a new PR https://github.com/saltstack/salt/pull/53385

**Pull request closed without merging**